File: WeeklyCode/week2/Add Two Numbers/chenhao.cpp
#include<iostream>
#include<algorithm>
#include<vector>
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode *addTwoNumbers(ListNode *l1, ListNode *l2) {
        ListNode vHead(0);
        ListNode* cur=&vHead;
        int carry=0;
        while(l1!=NULL || l2!=NULL){
            int value=carry;
            if(l1!=NULL){
                value+=l1->val;
                l1=l1->next;
            }
            if(l2!=NULL){
                value+=l2->val;
                l2=l2->next;
            }
            if(value>=10){
                value-=10;
                carry=1;
            }
            else{
                carry=0;
            }
            cur->next=new ListNode(value);
            cur=cur->next;
        }
        if(carry){
            cur->next=new ListNode(1);
        }
        return vHead.next;
    }
};

File: WeeklyCode/week6/Sort List/chenhao.py
class Solution:

    def merge(self,head1,len1,head2,len2):
        vHead=ListNode(0)
        cur=vHead
        p=q=0
        while p<len1 or q<len2:
            if p<len1 and ( q>=len2 or head1.val<head2.val ):
                p+=1
                cur.next=head1
                head1=head1.next
            else:
                q+=1
                cur.next=head2
                head2=head2.next
            cur=cur.next
        cur.next=None
        return vHead.next

    def sort0(self, length):
        if length==1:
            head=self.cur
            self.cur=self.cur.next
            return head
        lHead=self.sort0(length//2)
        rHead=self.sort0(length-length//2)
        return self.merge(lHead,length//2,rHead,length-length//2)

    # @param head, a ListNode
    # @return a ListNode
    def sortList(self, head):
        if head==None:
            return None
        length=0
        cur=head
        while cur!=None:
            length+=1
            cur=cur.next
        self.cur=head
        return self.sort0(length)